### PR TITLE
Chptr9 mailroom final

### PR DESCRIPTION
**Pull request by HawkeyeUW**
Final submission for UofW site

### CODE REVIEW

**Comment by HawkeyeUW** [comment]
Added Chapter 10 mailroom assignment to pull request.  Please grade both Chapters (9&10) since this is my final commit.  Additionally, my assignment for chapter 10 is the mailroom_fp.py file.

File: students/jvirtue/lesson09/mailroom.py [hunk 1]
```diff
@@ -0,0 +1,154 @@
...
+    
+    #Add Donations from Dict
+    def amount_add(self,donor_donation):
+        self.donations.append(float(donor_donation))
+
+    #Thank you message in console app
+    def thank_you(self, name, donation):
+        return "Thank you {} ".format(name) + "for the donation in the amount of {}!".format(donation)
+
+class Donor_Collections():
```
Comment (natasha-aleksandrova) [inline students/jvirtue/lesson09/mailroom.py]: no underscores in class names - should be CamelCase (`DonorCollection`)

File: students/jvirtue/lesson09/mailroom.py [hunk 2]
```diff
@@ -0,0 +1,154 @@
...
+
+    def donor_add(self,donor):
+        self.donors.append(donor)
+
+    #List out unique donors
+    def donor_list(self):
+        name_list = []
+        for donor in self._donors:
+            name_list.append(donor.name)
+        return name_list
```
Comment (natasha-aleksandrova) [inline students/jvirtue/lesson09/mailroom.py]: great place for a comprehension 😉 

`return [donor.name for donor in self._donors]`

File: students/jvirtue/lesson09/mailroom.py [hunk 3]
```diff
@@ -0,0 +1,154 @@
...
+                donor.donations.append(donation_amount)
+                break
+    
+    #Build list of donors, amount, count of contributions and average gift amount
+    def crunch_numbers(self):
+        summary_list = []
+        for donor in self._donors:
+            total = sum(donor.donations)
+            number_of_gifts = len(donor.donations)
+            average = total / number_of_gifts
```
Comment (natasha-aleksandrova) [inline students/jvirtue/lesson09/mailroom.py]: these should be provided by the `Donor` class:
```
donor.total_amount, donor.number_of_gifts, donor.average_amount
```

File: students/jvirtue/lesson09/mailroom.py [hunk 4]
```diff
@@ -0,0 +1,154 @@
...
+    def sort_key(self,crunch_numbers):
+        return crunch_numbers[1]
+
+    #Print report to console
+    def create_report(self):
+        print("\n","-" * 68,sep="")
+        print(f'{"Donor Name":<20}{"|":<1}{"Total Given":^15}{"|":<1}{"Num Gifts":^15}{"|":<1}{"Average Gift":>15}')
+        print("-" * 68)
+        sort_table = self.crunch_numbers()
+        [print(f'{item[0]:<20}{" ":<1}{"$":<1}{item[1]:>14}{" ":<1}{item[2]:>15}{" ":<1}{"$":<1}{item[3]:>14}') for item in sort_table]
```
Comment (natasha-aleksandrova) [inline students/jvirtue/lesson09/mailroom.py]: this is considered an anti-pattern, alternate code:
```
print("\n".join([<stuff here>]))
```

File: students/jvirtue/lesson09/mailroom.py [hunk 5]
```diff
@@ -0,0 +1,154 @@
...
+
+    #Thank you letter files to text file for all donors
+    def thanks_file(self):
+        for data in self.crunch_numbers():
+            filename = data[0].replace(" ", "_") + ".txt"
+            total_donation = data[1]
+            letter = ('Thank you {} for you generous contributions totaling {:.2f}!'.format(data[0], total_donation))
+            open(filename, 'w').write(letter)
```
Comment (natasha-aleksandrova) [inline students/jvirtue/lesson09/mailroom.py]: use `with` statement to ensure that file gets closed.

File: students/jvirtue/lesson09/mailroom.py [hunk 6]
```diff
@@ -0,0 +1,154 @@
...
+        print("-" * 68)
+        sort_table = self.crunch_numbers()
+        [print(f'{item[0]:<20}{" ":<1}{"$":<1}{item[1]:>14}{" ":<1}{item[2]:>15}{" ":<1}{"$":<1}{item[3]:>14}') for item in sort_table]
+
+    #Thank you letter files to text file for all donors
+    def thanks_file(self):
+        for data in self.crunch_numbers():
+            filename = data[0].replace(" ", "_") + ".txt"
+            total_donation = data[1]
+            letter = ('Thank you {} for you generous contributions totaling {:.2f}!'.format(data[0], total_donation))
```
Comment (natasha-aleksandrova) [inline students/jvirtue/lesson09/mailroom.py]: letter text should also be part of the Donor class

**Pull request merged**